android-agent/src/main/java/io/opentelemetry/android/internal/services/network/CarrierFinder.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.internal.services.network;
+
+import android.os.Build;
+import android.telephony.TelephonyManager;
+import androidx.annotation.RequiresApi;
+import io.opentelemetry.android.internal.services.network.data.Carrier;
+
+/**
+ * This class is internal and not for public use. Its APIs are unstable and can change at any time.
+ */
+@RequiresApi(api = Build.VERSION_CODES.P)
+public class CarrierFinder {
+
+    private final TelephonyManager telephonyManager;
+
+    public CarrierFinder(TelephonyManager telephonyManager) {
+        this.telephonyManager = telephonyManager;
+    }
+
+    public Carrier get() {
+        int id = telephonyManager.getSimCarrierId();
+        String name = null;
+        String mobileCountryCode = null;
+        String mobileNetworkCode = null;
+        String isoCountryCode = null;
```

Comment:
nit: should probably move these declarations closer to their usage.

Reply:
They are used in line 44 but they can change before reaching that line, that's why they're not close to line 44 because of the "ifs" needed to re-set them before that.